clientapi/auth/login.go

```diff
@@ -0,0 +1,83 @@
+// Copyright 2021 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package auth
+
+import (
+	"context"
+	"encoding/json"
+	"io"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/matrix-org/dendrite/clientapi/auth/authtypes"
+	"github.com/matrix-org/dendrite/clientapi/jsonerror"
+	"github.com/matrix-org/dendrite/setup/config"
+	uapi "github.com/matrix-org/dendrite/userapi/api"
+	"github.com/matrix-org/util"
+)
+
+// LoginFromJSONReader performs authentication given a login request body reader and
+// some context. It returns the basic login information and a cleanup function to be
+// called after authorization has completed, with the result of the authorization.
+// If the final return value is non-nil, an error occurred and the cleanup function
+// is nil.
+func LoginFromJSONReader(ctx context.Context, r io.Reader, accountDB AccountDatabase, userAPI UserInternalAPIForLogin, cfg *config.ClientAPI) (*Login, LoginCleanupFunc, *util.JSONResponse) {
```

Comment:
This is just to ensure cleanup after `completeAuth`. With more refactoring, that relationship could surely be inversed so that the cleanup isn't needed. I wanted to keep the diff small. Let me know if you want more.

Reply:
Some more rationale on this code: I'm trying to keep this function non-mutating. It's just parsing JSON and returning a struct. For tokens, we need to (1) parse, (2) create an access token, (3) remove the login token. And (3) shouldn't happen if (2) fails. However, I also wanted to avoid type-specific switches happening later in Login. Hence the idea of simply returning a cleanup function along with the `Login` (which is just the user and device identities).

The first alternative I can come up with is making `Login` an interface, and have type-specific implementations. That seemed overkill and a larger change, but certainly doable.